Feature/task-2

HanaNavratilova opened the pull request:
add contracts and database projects
create interface for list repository
move listItem to contracts project
add list repo interface implementation
add dependency resolver interface implementation
add container config using config in database project

Commit:
KA-325 add dummy DB layer
add contracts and database projects
create interface for list repository
move listItem to contracts project
add list repo interface implementation
add dependency resolver interface implementation
add container config using config in database project

Commit:
KA-325 change to 4.7.2 .new tramework

Commit:
KA-325 change appveyor config file

Commit:
KA-325 use NSubstitute in tests

Commit:
KA-325 add IBootstraper and UnityContainer RegisterTypesFrom extension method

Commit:
KA-325 change solution structure, add Api.Services project

Commit:
KA-325 add GetListItemLocation implementation and tests, update ServicesBootstraper Register method

Commit:
KA-325 add list Item comparer

Commit:
KA-325 remove unused usings

Commit:
KA-325 add correct namespaces

Commit:
KA-325 add ContainerConfig RegisterTypes test

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/App_Start/JsonSerializerConfig.cs:
```diff
@@ -3,7 +3,7 @@
+    public static class JsonSerializerConfig
     {
         public static void Register(HttpConfiguration config)
         {
```
Consider storing the `config.Formatters.JsonFormatter` in a local variable to access it in a shorter manner.

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/App_Start/ContainerConfig.cs:
```diff
@@ -0,0 +1,33 @@
+        }
+
+        internal static IUnityContainer RegisterTypesFrom<T>(this IUnityContainer container)
+            where T : IBootstraper, new()
```
Don't use just `T` - it is almost equivalent to single letter variables (I say almost since some information can usually be derived from a single-letter variable name, while here it only means "type" and that is obvious, bringing literally no value).

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/App_Start/ContainerConfig.cs:
```diff
@@ -0,0 +1,33 @@
+        }
+
+        internal static void RegisterTypes(IUnityContainer container)
+        {
```
Bodyless function? 😛

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/DependencyResolver.cs:
```diff
@@ -0,0 +1,53 @@
+{
+    public class DependencyResolver : IDependencyResolver
+    {
+        protected IUnityContainer _container;
```
I'm suggested to rename the field to `Container` – it is caused by the lack of R# settings in your repository, please, read a bit more on [R# configuration options](https://www.jetbrains.com/help/resharper/Sharing_Configuration_Options.html) add team-wide settings that would comply with your code-style (or our coding rules, really)…

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/DependencyResolver.cs:
```diff
@@ -0,0 +1,53 @@
+            }
+            catch (ResolutionFailedException)
+            {
+                return new List<object>();
```
Definitely use `Enumerable.Empty` here, please.

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/DependencyResolver.cs:
```diff
@@ -0,0 +1,53 @@
+            {
+                return _container.Resolve(serviceType);
+            }
+            catch (ResolutionFailedException)
```
Please, have a short debug session with this catch - I know you have a test for registered dependencies, but you should still let your code [fail fast](https://dzone.com/articles/fail-fast-principle-in-software-development) in case some dependency is missing. By returning `null` (or empty collection below) you actually mask where an error originates and mask root cause. You will find that there is a reason why the boilerplate looks like this, however, it would be much better to throw `ResolutionFailedException` when a type is forgotten to be registered…

I will insist on this in the next task, consider postponing it until you separate `UnityContainer`…

HanaNavratilova replied:
https://stackoverflow.com/questions/3819842/catch-exception-with-condition

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api/MyPerfectOnboarding.Api.csproj:
```diff
@@ -0,0 +1,204 @@
+      <DependentUpon>Global.asax</DependentUpon>
+    </Compile>
+    <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="UrlLocatorConfig.cs" />
```
I thought this file location and name were just temporary 😆 
1. You actually configure route names and you should not obscure this fact by a service name - there might be more services consuming this object (well, not in this onboarding, but still) [consider renaming interface as well 😉]
1. I wouldn't be afraid to create a `Configuration` folder or something like that... Keep the project root clear for the bootstrapping/entry-point code (so it can be easily located)

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api.Services/ServicesBootstraper.cs:
```diff
@@ -0,0 +1,18 @@
+        public IUnityContainer RegisterTypesTo(IUnityContainer container)
+            => container
+                .RegisterType<IUrlLocator, UrlLocator>(new HierarchicalLifetimeManager())
+                .RegisterType<HttpRequestMessage>(new InjectionFactory(unityContainer => HttpContext.Current.Items["MS_HttpRequestMessage"]));
```
What a nice, well-named private static function would this lambda form? 😛

PetrSvirak replied:
You miss `LifetimeManager` here 😉

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Contracts/Database/IListRepository.cs:
```diff
@@ -0,0 +1,19 @@
+
+        Task DeleteItemAsync(Guid id);
+
+        Task EditItemAsync(Guid id, ListItem editedItem);
```
Why pass `id` when your - now surely well-formed - `editedItem` already contains one? 😛

PetrSvirak replied:
I wouldn't be afraid to call the method `replace` BTW…

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Contracts/Database/IListRepository.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        Task<ListItem> AddItemAsync(ListItem item);
+
+        Task DeleteItemAsync(Guid id);
```
👍

PetrSvirak replied:
Not that it is necessary, however, have a look at the other allowed response for REST delete - if you wanted to return the deleted item (more of a politeness than necessity), your repository won't allow you to. Consider making the repository contract broader while feeling free to left controller untouched (same logic can be applied to `EditItem`, btw)

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Contracts/Models/ListItem.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public override string ToString()
+        {
+            return $"{nameof(Id)}: {Id}, {nameof(Text)}: {Text}, {nameof(IsActive)}: {IsActive}, {nameof(CreationTime)}: {CreationTime}, {nameof(LastUpdateTime)}: {LastUpdateTime}";
```
R# underlines this one 😛

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Contracts/Api/IUrlLocatorConfig.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace MyPerfectOnboarding.Contracts.Api
```
What about placing this interface closer to the `IUrlLocator`? I mean, it can stay where it is, however, wouldn't it be clearer to group interfaces by the domain where applicable (you need this one for `api.services`, `location` more precisely, the place where implementation resides is not that relevant in my eyes)?

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Database/ListRepository.cs:
```diff
@@ -0,0 +1,39 @@
+using System.Threading.Tasks;
+using MyPerfectOnboarding.Contracts.Database;
+
+namespace MyPerfectOnboarding.Database
```
Place the file into a folder, please - I know it might result in a namespace like `MyPerfectOnboarding.Database.Database`  or `MyPerfectOnboarding.Database.Repository`, however, it states your project structure more clearly and does not obscure bootstrapping code…

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Services.Tests/Location/UrlLocationTests.cs:
```diff
@@ -0,0 +1,39 @@
+            urlHelper.Link(name, Arg.Is<object>(obj => IsGivenIdInObject(obj, id))).Returns(expectedUri.ToString());
+            var urlLocatorConfig = Substitute.For<IUrlLocatorConfig>();
+            urlLocatorConfig.ListItemRouteName.Returns(name);
+            var urlLocation = new UrlLocator(urlHelper, urlLocatorConfig);
```
Please, try introducing a private method that would build the `UrlLocator` (and its dependencies) - I believe it will improve the readability of the test (since `arrange` part is quite large now).

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Services.Tests/Location/UrlLocationTests.cs:
```diff
@@ -0,0 +1,39 @@
+            var expectedUri = new Uri($"http://www.aaa.com/try/to/create/uri/with/id/{id}/and/it/is/awesome");
+            const string name = "aaaaa";
+            var urlHelper = Substitute.For<UrlHelper>();
+            urlHelper.Link(name, Arg.Is<object>(obj => IsGivenIdInObject(obj, id))).Returns(expectedUri.ToString());
```
It's not `obj`:
1. it's unnecessary abbrev. 😛
2. that is not describing the nature of the second argument of the `Link` method (I mean, have a look at the `Link` method itself, they didn't call their second parameter "object" 😝)

PetrSvirak replied:
Also, wrap commands by the dot and consider placing parameters on their own lines, please. 

What do you think about:
```
var urlHelper = Substitute.For<UrlHelper>();
urlHelper
    .Link(
        routeName,
        Arg.Is<object>(obj => IsGivenIdInObject(obj, itemId)))
    .Returns(expectedUri.ToString());

var urlLocatorConfig = Substitute.For<IUrlLocatorConfig>();
urlLocatorConfig
    .ListItemRouteName
    .Returns(routeName);
```
?

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/App_Start/ContainerConfigTests.cs:
```diff
@@ -0,0 +1,51 @@
+            typeof(IBootstraper)
+        };
+
+        private static readonly HashSet<Type> InterfacesToRegister = new HashSet<Type>
```
I would prefer arrays here - you never change collections at runtime and if you happen to include a type twice, you won't be notified anyway. Later on, you work with collections through the Enumerable interface (what is correct IMO), thus you won't leverage any of `HasSet`'s aspects… Declaring collections in the following manner seems a bit more robust and clear to me:
```
private static readonly IEnumerable<Type> X = new []
```
What do you think?

PetrSvirak replied:
Maybe the name is a bit misleading. These are the type you don't own (don't have them in your code), so maybe `ExplicitelyRegisteredTypes` (or `TypesToRegisterExplicitly`) or `ForeignTypes` or something similar might give a better picture on the collection content…

PetrSvirak replied:
Also, reconsider the name `interfaces`, they mustn't be - it might be `abstract classes` or even `classes`. Since `interface` is a reserved word, don't use it for something that is not necessarily an interface (you might have found `ClassesToRegister` field just below in about a week or two this way 😆)

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/App_Start/ContainerConfigTests.cs:
```diff
@@ -0,0 +1,51 @@
+        private static readonly HashSet<Type> InterfacesToRegister = new HashSet<Type>
+        {
+            typeof(HttpRequestMessage),
+            typeof(IUnityContainer)
```
I would normally ask you to implement fake `IUnityContainer` to avoid this "side-effect" of using a real container in your test (it renders it kind of `non-unit` in my eyes), but since you will change this in next task anyway, let there be `UnityContainer` for now…

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/App_Start/ContainerConfigTests.cs:
```diff
@@ -0,0 +1,51 @@
+
+            
+            ContainerConfig.RegisterTypes(container);
+            var interfaces = container.Registrations.Select(r => r.RegisteredType).ToHashSet();
```
Consider building two collection here: One with missing registrations and one with unwanted/unexpected registration. You can then assert these 2 collections to be empty. It has 2 benefits: one is that you express more explicitly that the `expectedInterfaces` and `interfaces` sets should be the same, the other is that assert message only states extra interfaces, not 2 (potentially large) sets that differ by 1 (kinda long) string item…

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/App_Start/ContainerConfigTests.cs:
```diff
@@ -0,0 +1,51 @@
+                .Except(InterfacesNotToRegister)
+                .ToHashSet();
+
+            
```
One too many empty lines here 😉

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+        private IListRepository _repository;
+        private IUrlLocator _location;
+
+       [SetUp]
```
Your attribute is not properly aligned 😉

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+            _repository.GetAllItemsAsync().Returns(Task.FromResult(_items));
+
+            var message = await _listController.ExecuteAction(controller => controller.GetAsync());
+
```
You usually don't place an empty line here (which is IMO correct) 😉

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+            message.TryGetContentValue(out IEnumerable<ListItem> items);
+
+            Assert.That(message.StatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(items, Is.EqualTo(_items).Using(new ListItemEqualityComparer()));
```
You repeat this code all over the place: ``.Using(new ListItemEqualityComparer())``.
1. You can have a single `ListItemEqualityComparer` since Your implementation of equality is a pure function
1. You can have a single extension method so when you decided to use a different comparer, it will be easy to replace

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+        [Test]
+        public async Task Post_CreatedReturned()
+        {
+            var item = new ListItem {Text = "newItem"};
```
Let's use an adjective here to describe what `item` this one is…

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+        public async Task Post_CreatedReturned()
+        {
+            var item = new ListItem {Text = "newItem"};
+            var newItem = new ListItem
```
Even consider renaming this one in regards to the one on the line above…

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+            var message = await _listController.ExecuteAction(controller => controller.PostAsync(item));
+
+            Assert.That(message.StatusCode, Is.EqualTo(HttpStatusCode.Created));
+            Assert.That(message.Headers.Location, Is.EqualTo(expectedUri));
```
There is an assertion missing, isn't it? You should check what item was returned…

PetrSvirak replied:
Also, I believe this one to be the most significant example: 

Imagine this test is failing because of your mocked repository returns `null` (there is an error in your implementation in this imaginary scenario and it is correct that the test fails). Now, your status code is OK, but `expectedUrl` is obviously not and the test fails with this assertion. You will not see that item returned in response body is missing and might start investigating code that calls `location assembler` instead of the one that calls repository. 

Have a look at [`Assert.Multiple`](https://github.com/nunit/docs/wiki/Multiple-Asserts), please.

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,128 @@
+        [Test]
+        public async Task Put_NoContentReturned()
+        {
+            var id = Guid.NewGuid();
```
This kind of renders text unrepeatable, please, use fixed GUID.

PetrSvirak reviewed (changes requested):
Perfect work with commits - there is but few of them and they are reasonably named. I really like your "commit-work". They contain what is not described in commit message from time to time, but in general, they are really nice. 

The very same thing can be said about your solution. I found quite a lot of nitpicking details you have fixed in no-time and one or two slightly bigger details I want you to address; but otherwise, it's your perfect onboarding.

Let's fix these remaining simpletons and start on the next task! 😃

Commit:
KA-325 code refactoring
- store config.Formatters.JsonFormatter as a local variable
- rename T type in containerConfig
- rename _container to Container in DependencyResolver
- delete unused compile include in Api.csproj
- add missing LifetimeManager in ServicesBootstraper
- fixup ListRepository interface
- fixup namespaces
- changins structure in UrlLocationTests, ContainerConfigTests
- refactor ListControllerTests (renaming, deleting unused things, ...)

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/App_Start/ContainerConfigTests.cs:
```diff
@@ -0,0 +1,57 @@
+            var expectedTypes = GetExpectedTypes();
+
+            ContainerConfig.RegisterTypes(container);
+            var types = container.Registrations.Select(r => r.RegisteredType).ToArray();
```
The `types` could be called `registeredTypes` 😛

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Services.Tests/Location/UrlLocationTests.cs:
```diff
@@ -0,0 +1,55 @@
+            return new UrlLocator(urlHelper, urlLocatorConfig);
+        }
+
+        private static bool IsGivenIdInRouteValues(object obj, Guid expectedId)
```
Did you know, you can do this in the new C#:
```
private static Func<object, bool> IsGivenIdInRouteValues(Guid expectedId)
    => @object =>
    {
        // ReSharper disable once CollectionNeverUpdated.Local
        var dictionary = new RouteValueDictionary(@object);
        var id = (Guid) dictionary["id"];

        return id == expectedId;
    };
```

Besides some familiarity with action creator factories from `JavaScript` tasks, it also allows you to write substitute's code like this:
```
urlHelper
    .Link(
        routeName,
        Arg.Is<object>(IsGivenIdInRouteValues(id)))
    .Returns(expectedUri.ToString());
```

Isn't it a nasty piece of code? 😆

PetrSvirak replied:
Unfortunately, the test will stop passing since once the `Is` method is provided with an object - a reference, really - it tries to match it against provided argument instead of invoking it, so to make the test work, one would need to us `Is` like this: `Arg.Is<object>(o => IsGivenIdInRouteValues(id)(o)))` – that is not so cool 😢

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,134 @@
+            message.TryGetContentValue(out IEnumerable<ListItem> items);
+
+            Assert.That(message.StatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(items, Is.EqualTo(_items).Using(_listItemEqualityComparer));
```
I expected you to combine these two: https://github.com/HanaNavratilova/kentico-onboarding-cs/pull/3#discussion_r219792366 😆 

Anyway, it is true that you now repeat `.Using(_listItemEqualityComparer)` in this controller and You would surely be tempted to create another instance of the `ListItemEqualityComparer` were there another controller to be tested and your implementation is not `Lazy` (ask somebody why; I'm sure Juraj or Linda will know this time! 🤣).

However, I can wait with this detail until you finish the next task - since it now suffices.

PetrSvirak commented on MyPerfectOnboardingApplication/Tests/MyPerfectOnboarding.Api.Tests/Controllers/ListControllerTests.cs:
```diff
@@ -0,0 +1,134 @@
+                Text = "newItem",
+                IsActive = false,
+                CreationTime = new DateTime(1589, 12, 3),
+                LastUpdateTime = new DateTime(1896, 4, 7)
```
I love the dates 🤣 I can see that completing some items may sometimes take much longer than others 😆

PetrSvirak commented on MyPerfectOnboardingApplication/Sources/MyPerfectOnboarding.Api.Services/ServicesBootstraper.cs:
```diff
@@ -0,0 +1,18 @@
+        public IUnityContainer RegisterTypesTo(IUnityContainer container)
+            => container
+                .RegisterType<IUrlLocator, UrlLocator>(new HierarchicalLifetimeManager())
+                .RegisterType<HttpRequestMessage>(new HierarchicalLifetimeManager(), new InjectionFactory(unityContainer => HttpContext.Current.Items["MS_HttpRequestMessage"]));
```
You kind of forget about this one: https://github.com/HanaNavratilova/kentico-onboarding-cs/pull/3#discussion_r219759984 😜

PetrSvirak reviewed (approved):
Devil dwells in details 😈 

You seem to forget about one comment, we misunderstood each other in one case, there is (at least) a one place for naming improvement and I seem to forget to mention to You that `WebApiConfig` should now be named `RoutesConfig` (since it only cares about routes, really) - but that is it. 

Hereby I render this task approved 🎆 🍾 🎺
